sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/LockRenewSubscriber.java

```diff
@@ -0,0 +1,108 @@
+    private final CoreSubscriber<? super ServiceBusReceivedMessage> actual;
+
+    LockRenewSubscriber(CoreSubscriber<? super ServiceBusReceivedMessage> actual, Duration maxAutoLockRenewDuration,
+        LockContainer<LockRenewalOperation> messageLockContainer,
+        Function<String, Mono<OffsetDateTime>> onRenewLock) {
+        this.onRenewLock = Objects.requireNonNull(onRenewLock, "'onRenewLock' cannot be null.");
+        this.actual = actual;
+        this.messageLockContainer = Objects.requireNonNull(messageLockContainer,
+            "'messageLockContainer' cannot be null.");
+        Objects.requireNonNull(maxAutoLockRenewDuration, "'maxAutoLockRenewDuration' cannot be null.");
+        this.maxAutoLockRenewal = maxAutoLockRenewDuration;
+    }
+    /**
+     * On an initial subscription, will take the first work item, and request that amount of work for it.
+     * @param subscription Subscription for upstream.
+     */
+    @Override
+    protected void hookOnSubscribe(Subscription subscription) {
+        Objects.requireNonNull(subscription, "'subscription' cannot be null.");
+        actual.onSubscribe(subscription);
+    }
+
+    /**
+     * When upstream has completed emitting messages.
+     */
+    @Override
+    public void hookOnComplete() {
+        logger.info("Upstream has completed.");
+        actual.onComplete();
+    }
+
+    @Override
+    protected void hookOnError(Throwable throwable) {
+        logger.error("Errors occurred upstream.", throwable);
+        actual.onError(throwable);
+        dispose();
+    }
+
+    @Override
+    protected void hookOnNext(ServiceBusReceivedMessage message) {
+        final String lockToken = message.getLockToken();
+        final OffsetDateTime lockedUntil = message.getLockedUntil();
+        final LockRenewalOperation renewOperation;
+
+        if (!Objects.isNull(lockedUntil) && !Objects.isNull(lockToken)) {
+
+            renewOperation = new LockRenewalOperation(lockToken, maxAutoLockRenewal, false, onRenewLock,
+                lockedUntil, message::setLockedUntil);
+
+            try {
+                messageLockContainer.addOrUpdate(lockToken, OffsetDateTime.now().plus(maxAutoLockRenewal),
+                    renewOperation);
+            } catch (Exception e) {
+                logger.error("Exception occurred while updating lockContainer for token [{}].", lockToken, e);
+                onError(e);
+            }
+        } else {
+            logger.warning("Unexpected, LockToken [{}] or lockedUntil [{}] is not present for message [{}].",
+                lockToken, lockedUntil, message.getSequenceNumber());
+            renewOperation = COMPLETED_ONE;
+        }
+
+        try {
+            actual.onNext(message);
+        } catch (Exception e) {
+            logger.error("Exception occurred while handling downstream onNext operation.", e);
+            onError(e);
+        } finally {
+            renewOperation.close();
```

Comment:
When the renewOperation is closed, should we also remove this from `messageLockContainer`?